Add SSR guide (#110)
radix-ui/website: pages/primitives/docs/overview/server-side-rendering.mdx

navRank: 6
---

## Overview

Server side rendering or `SSR`, is a technique used to render components to HTML on the server, as opposed to rendering them only on the client.

Static rendering is another similar approach. Instead it pre-renders pages to HTML at build time rather than on each request.

You should be able to use all of our primitives with both approaches, for example with [Next.js](https://nextjs.org/) or [Gatbsy](https://www.gatsbyjs.com/), with very little setup required.

## IdProvider

The way `SSR` works with React is by rendering components to HTML on the server and then hydrating the HTML, thus enhancing it with interactivity.

For all of this to function properly, React has to match the DOM structure between client and server. This is potentially an issue for Radix Primitives because we rely on generating random `id`s internally so that accessibility works out of the box.

To enable more deterministic `id`s between server and client and ensure that no mismatch occurs, wrap your app with our `IdProvider` like so:

```tsx
import { IdProvider } from '@radix-ui/react-id';
export default () => (
  <IdProvider>
    <App />
  </IdProvider>

> );

Reviewer: Ah bum, looks like my `suggestion` broke the closing backticks here 🙈

---
title: Server side rendering
description: Radix Primitives can be rendered on the server. However a little setup is required to make sure they work properly.
navRank: 6
---

## Overview

Server side rendering or `SSR`, is a technique used to render components to HTML on the server, as opposed to rendering them only on the client.

Static rendering is another similar approach. Instead it pre-renders pages to HTML at build time rather than on each request.

You should be able to use all of our primitives with both approaches, for example with [Next.js](https://nextjs.org/) or [Gatbsy](https://www.gatsbyjs.com/), with very little setup required.

## IdProvider

The way `SSR` works with React is by rendering components to HTML on the server and then hydrating the HTML, thus enhancing it with interactivity.

For all of this to function properly, React has to match the DOM structure between client and server. This is potentially an issue for Radix Primitives because we rely on generating random `id`s internally so that accessibility works out of the box.

> You will need to wrap your app with an `IdProvider` like so:
> 
> ```tsx
> import { IdProvider } from '@radix-ui/react-id';
> 
> export default () => (
>   <IdProvider>
>     <App />
>   </IdProvider>
> );
> ```
> 
> This will enable more deterministic `id`s between server and client and ensure that no mismatch will happen.

Reviewer: Can we move this part above please? I find that reasoning first then the how after reads better 🙏 

```suggestion
To enable more deterministic `id`s between server and client and ensure that no mismatch occurs, wrap your app with our `IdProvider` like so:

```tsx
import { IdProvider } from '@radix-ui/react-id';
export default () => (
  <IdProvider>
    <App />
  </IdProvider>
);
```
```
